java/ql/src/semmle/code/java/dataflow/FlowSteps.qll

```diff
@@ -20,6 +20,60 @@ private module Frameworks {
   private import semmle.code.java.frameworks.ApacheHttp
 }
 
+/**
+ * A method that returns the exact value of one of its parameters or the qualifier.
+ *
+ * Extend this class and override `returnsValue` to add additional value-preserving steps through a
+ * method that should be added to the basic local flow step relation.
+ *
+ * These steps will be visible for all global data-flow purposes, as well as via
+ * `DataFlow::Node.getASuccessor` and other related functions exposing intraprocedural dataflow.
+ */
+abstract class ValuePreservingMethod extends Method {
+  /**
+   * Holds if this method returns precisely the value passed into argument `arg`.
+   * `arg` is a parameter index, or is -1 to indicate the qualifier.
+   */
+  abstract predicate returnsValue(int arg);
+}
+
+/**
+ * A method that returns the exact value of its qualifier (e.g., `return this;`)
+ *
+ * Extend this class and override `returnsValue` to add additional value-preserving steps through a
+ * method that should be added to the basic local flow step relation.
+ *
+ * These steps will be visible for all global data-flow purposes, as well as via
+ * `DataFlow::Node.getASuccessor` and other related functions exposing intraprocedural dataflow.
+ */
+abstract class FluentMethod extends ValuePreservingCallable {
+  override predicate returnsValue(int arg) { arg = -1 }
+}
+
+private class StandardLibraryValuePreservingCallable extends ValuePreservingCallable {
+  int returnsArgNo;
+
+  StandardLibraryValuePreservingCallable() {
+    this.getDeclaringType().hasQualifiedName("java.util", "Objects") and
+    (
+      this.hasName(["requireNonNull", "requireNonNullElseGet"]) and returnsArgNo = 0
+      or
+      this.hasName("requireNonNullElse") and returnsArgNo = [0 .. this.getNumberOfParameters() - 1]
+      or
+      this.hasName("toString") and returnsArgNo = 1
+    )
+    or
+    this.getDeclaringType()
+        .getSourceDeclaration()
```

Comment:
We should possibly get rid of this line and move the responsibility for handling generics into `DataFlowUtil.qll`.